Enhancements to WidgetSelector, RxCommandBuilder. New executable widget.

AlexBacich opened the pull request:
Hi, thanks for awesome library. My PR with enhancements. Please let me know what do you think about them. If you like them I have more ideas to implement and cleanup.

WidgetSelector
- Made onTrue, onFalse optional. Blank widget is used if null. This is usable if you need to put custom widget only for one branch

RxCommandBuilder
- Made stateless widget
- Made possible to build with blank of Placeholder and Error. Blank widgets used in that case
- Made possible to build with blank Busy. Also used platform parameter to build progress indicator if busy is null

RxCommandExecutable
- New widget. Lets you wrap your widget and give functions which will be run once new data is emitted in streams. This way you may keep your widgets stateless and clean.

Commit:
WidgetSelector
- Made onTrue, onFalse optional. Blank widget is used if null. This is usable if you need to put custom widget only for one branch

RxCommandBuilder
- Made stateless widget
- Made possible to build with blank of Placeholder and Error. Blank widgets used in that case
- Made possible to build with blank Busy. Also used platform parameter to build progress indicator if busy is null

RxCommandExecutable
- New widget. Lets you wrap your widget and give functions which will be run once new data is emitted in streams. This way you may keep your widgets stateless and clean.

escamoteur commented on lib/src/rx_command_executable.dart:
```diff
@@ -0,0 +1,83 @@
+import 'package:flutter/widgets.dart';
+import 'package:rx_command/rx_command.dart';
+
+/// Wrap your [Widget] into [RxCommandExecutable] widget to get an ability to listen for events
```
I'm not clear what you would do with such a widget honestly. I think if you need something like that you probably also need a stateful widget and register your own CommandListener. With this widget you can't pass any values to the build method as its fixed.
Perhaps you can share your thoughts on this with me.

escamoteur reviewed (commented):
Looks good. I made one comment on your new widget. If you have more Ideas, I'm happy to add you as a maintainer as this is one of my step children packages.

AlexBacich commented:
Exactly. This RxCommandExecutable widget creates CommandListener under the hood and takes care of creating it, listening and disposing. Its purpose is to make CommandListener easier in use. 

Let me give example: 


```
class SomeStatelessClass extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return RxCommandExecutable(
        command: bloc.register,
        onError: (error) => BotToast.showText(text: 'Some registration error'),
        child: SomeChildWidget(....),
    );
  }
}
```

Once this widget goes away from render tree - its child RxCommandExecutable will take care of unsubscribing and disposing from CommandListener.

AlexBacich commented:
And thanks for suggesting to be maintainer! I adore this package and will post more widgets soon. Working on RxTextEdit at the moment. Again, its purpose to hide all controllers into it and instead provide Stream based interfaces to use it.

escamoteur commented:
> Exactly. This RxCommandExecutable widget creates CommandListener under the hood and takes care of creating it, listening and disposing. Its purpose is to make CommandListener easier in use.
> 
> Let me give example:
> 
> ```
> class SomeStatelessClass extends StatelessWidget {
>   @override
>   Widget build(BuildContext context) {
>     return RxCommandExecutable(
>         command: bloc.register,
>         onError: (error) => BotToast.showText(text: 'Some registration error'),
>         child: SomeChildWidget(....),
>     );
>   }
> }
> ```
> 
> Once this widget goes away from render tree - its child RxCommandExecutable will take care of unsubscribing and disposing from CommandListener.

I'm not sure how often you will need that compared to a CommandBuilder. 
I would like to give it another name something like RxCommandHandler. Executable doesn't really is a good fit.

AlexBacich commented:
I would agree on naming - RxCommandHandler sounds good. 

As for usage - I'm using a lot when command result has to init some UI action requiring context. Showing snackbar, toast, dialog, etc. All of improvements and this new widget - result of practical usage in my projects, not vice versa :)

escamoteur commented:
ah, Ok I see. ever thought that maybe a mixin would be the better solution for this?

escamoteur commented:
are you adding the renaming to this PR?

AlexBacich commented:
Yes, later today or tomorrow I will add naming change and will add information in readme.md to be consistent. 
As for using mixins - haven't used them a lot but curious to see how to implement it. Can you hint on how to implement it using mixins? 

I guess we should declare widgets as:
MyWidget extends StatelessWidget with RxCommandHandler...
But then how to pass command and listening functions to mixin?

AlexBacich commented:
@escamoteur , any hint for Mixin inplementation? How you see it should working, I'll deal with details.

escamoteur commented:
you could give the mixin a `listenToCommand` function where you pass everything over?

AlexBacich commented:
@escamoteur , is your idea to add mixin to any Stateless \ Stateful class with listenToCommand function with the command itself and actions as parameters? That is doable, but primary problem would be that mixin won't be able to get notification from its hosting widget that its lifecycle is over and time to stop listening for events \ dispose itself. I would need to add another method dispose and that again would force user to make widget Stateful. 

Or did you have anything else in your mind? 

Just so we get back to track to this - I can finalize MD and merge with what I did so far. 

Let me know what do you think, thanks!

escamoteur commented:
have a look how I do it in the GetIt mixin: https://github.com/escamoteur/get_it_mixin

AlexBacich commented:
GetIt implementation itself is dark forest so far, but I understood key idea.

So you are getting stateless widget lifecycle through its element. 

```
@override
  StatelessElement createElement() => _StatelessMixInElement(this);
```

It's interesting, I will work on it and create mixin for CommandHandler. 

p.s. if you have time - one question about mix_in.dart.
`class _StatelessMixInElement<W extends GetItMixin> extends StatelessElement with _GetItElement`
Why did you mixin here _GetItElement instead of adding field directly inside _StatelessMixInElement? 
`_MixinState _state;`

escamoteur commented:
because I wanted to reuse all the state logic for the Stateful version of the mixin.

Commit:
- Added basic RxCommandHandlerMixin

AlexBacich commented:
@escamoteur , I've done basic implementation for RxCommandHandlerMixin. So far for StatelessWidget > onValue only.

Stumbled here since it's not working when I run in release mode. But works perfectly in debug mode. If I move out _command outside of CommandWidget - it starts working in release too. 

But this approach worries that It allows overriding any methods from StatelessElement. User is forced to start overriding and see all of them.

Commit:
- Changed mixin to override single value: commandListener

AlexBacich commented:
Added another commit to work with only override value where user can put entire CommandListener. Is this your initial idea on how it should work? 

Please provide some feedback and I'll finalize class.

escamoteur commented:
Hi Alex,

why do you think the user has to override Element functions?
I would pass the command to the `listen()` function. Have a look at my `registerStreamHandler()` function of the GetIt mixin. In your case the user would have to pass the command. you can take the get_it_mixin code and adapt it to your needs.

AlexBacich commented:
I don't think he would, it's just about seeing things he shouldn't see. 

Looked at the `registerStreamHandler()` - like it, will go with it.

Commit:
- Added RxCommandHandlerMixin
- Removed RxCommandExecutor
- Added ReactiveWidget (similar to ReactiveBuilder but without context in methods)

AlexBacich commented:
Ready to push. 

- Added RxCommandHandlerMixin
- Added ReactiveWidget. This is new one. It's similar to ReactiveBuilder but without context in methods. I'm getting much use of it. 
- Updated readme.md with two new widgets. 

======

I've decided not to go with `registerStreamHandler()`. It has two disadvantages to my mind. 
First - you can add mixin to your class and forget calling `register..()`. With my approach its necessary getter field. 
Second - I'm not expecting to put anything executable in build() method. Even Flutter documentation states this: `This method can potentially be called in every frame and should not have any side effects beyond building a widget.`

You gave me rights to merge, but still this is my first commit to this repository - please review before I do merge.

AlexBacich commented:
@escamoteur, can I push?

escamoteur commented on README.md:
```diff
@@ -292,4 +290,96 @@ class Example extends StatelessWidget {
+  }
+
+  @override
+  RxCommandListener get commandListener => RxCommandListener<void, DateTime>(
```
Sorry I hadn't time to look at this earlier.
Why did you choose this way with the override instead to give the mixin a `registerCommandHandler()` that has all the parameters a listener has? This way you always can only listen to one Command ins a Widget.

AlexBacich replied:
Posted reason for that above. It can be Iterable of listeners then. 

I've decided not to go with `registerStreamHandler()`. It has two disadvantages to my mind. 
First - you can add mixin to your class and forget calling `register..()`. With my approach its necessary getter field. 
Second - I'm not expecting to put anything executable in build() method. Even Flutter documentation states this: `This method can potentially be called in every frame and should not have any side effects beyond building a widget.`

escamoteur replied:
oh sorry I missed your comment. sure that is a valid point. although the thing with executing something in the build method isn't that critical. I will merge this now
